From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sat, 29 Jul 2023 18:53:58 +0200
Subject: [PATCH] Client perspective - Block state visualization - Chunk packet
 data can use buffer partially

Instead of assuming we can precisely pre-calculate the size of the byte buffer needed
to contain the data in ClientboundLevelChunkPacketData, we begin by assuming that the
used buffer length is indeed the size that we pre-calculated, but we store it in a field (#usedBufferLength)
that may be modified if it later becomes clear the the buffer is in fact not entirely filled.
This gives us the opportunity to perform further changes to the chunk data after it has already been
stored in the buffer, when those changes may have a higher final buffer length than we initially
calculated. In that case, we can, instead of allocating a buffer with the precise length known in advance,
allocate a buffer with a length that is sufficient instead, and later set #usedBufferLength to the length
that was actually used, without having to re-allocate a new byte buffer with the correct trimmed size.

License: Fiddle Public License Agreement 1.0 (included in license/FPL-1.0.txt)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 3095ecfc7f6aa95284c7812e52f5fb68d65c9061..28cb41ebc0c1a71681ac656bba2834a39d1fe18f 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -24,6 +24,7 @@ public class ClientboundLevelChunkPacketData {
     private static final int TWO_MEGABYTES = 2097152;
     private final CompoundTag heightmaps;
     private final byte[] buffer;
+    public int bufferUsedLength; // Fiddle - client perspective - block state visualization - chunk packet data can use buffer partially
     private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
     // Paper start
     private final java.util.List<net.minecraft.network.protocol.Packet> extraPackets = new java.util.ArrayList<>();
@@ -46,6 +47,7 @@ public class ClientboundLevelChunkPacketData {
         }
 
         this.buffer = new byte[calculateChunkSize(chunk)];
+        this.bufferUsedLength = this.buffer.length; // Fiddle - client perspective - block state visualization - chunk packet data can use buffer partially - initially we assume the entire buffer will be used
 
         // Paper start - Anti-Xray - Add chunk packet info
         if (chunkPacketInfo != null) {
@@ -82,6 +84,7 @@ public class ClientboundLevelChunkPacketData {
                 throw new RuntimeException("Chunk Packet trying to allocate too much memory on read.");
             } else {
                 this.buffer = new byte[i];
+                this.bufferUsedLength = this.buffer.length; // Fiddle - client perspective - block state visualization - chunk packet data can use buffer partially - initially we assume the entire buffer will be used
                 buf.readBytes(this.buffer);
                 this.blockEntitiesData = buf.readList(ClientboundLevelChunkPacketData.BlockEntityInfo::new);
             }
@@ -90,8 +93,10 @@ public class ClientboundLevelChunkPacketData {
 
     public void write(FriendlyByteBuf buf) {
         buf.writeNbt(this.heightmaps, DeepReplacementScope.NOTHING); // Fiddle - client perspective - deep replacements - pass scope - to FriendlyByteBuf#writeNbt - this never contains replaceables
-        buf.writeVarInt(this.buffer.length);
-        buf.writeBytes(this.buffer);
+        // Fiddle start - client perspective - block state visualization - chunk packet data can use buffer partially
+        buf.writeVarInt(this.bufferUsedLength);
+        buf.writeBytes(this.buffer, 0, this.bufferUsedLength);
+        // Fiddle end - client perspective - block state visualization - chunk packet data can use buffer partially
         buf.writeCollection(this.blockEntitiesData, (buf2, entry) -> {
             entry.write(buf2);
         });
